src/txn/txn_ckpt.c

```diff
@@ -189,10 +211,19 @@ __checkpoint_apply(WT_SESSION_IMPL *session, const char *cfg[],
 			continue;
 		WT_WITH_DHANDLE(session, session->ckpt_handle[i],
 		    ret = (*op)(session, cfg));
-		WT_RET(ret);
+		WT_ERR(ret);
 	}
 
-	return (0);
+err:
+	/*
+	 * If we have an error somewhere in processing the handles, then
+	 * we need to mark earlier trees dirty.  Recursively call the apply
+	 * function to reset on error.
+	 */
+	if (ret != 0)
+		(void)__checkpoint_apply(session, cfg, __checkpoint_err_reset);
```

Comment:
One note, this code (like the original), marks all of the handles dirty, not just the handles we've checkpointed. I think that's correct, and I don't see any point in optimizing the loop to avoid it, but wanted to call it out.